File: android/src/test/kotlin/com/bbflight/background_downloader/AuthTest.kt
package com.bbflight.background_downloader

import org.junit.Test
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue

class AuthTest {

    private fun createAuth(
        accessToken: String? = null,
        accessHeaders: Map<String, String> = emptyMap(),
        accessQueryParams: Map<String, String> = emptyMap(),
        accessTokenExpiryTime: Long? = null,
        refreshToken: String? = null,
        refreshHeaders: Map<String, String> = emptyMap(),
        refreshQueryParams: Map<String, String> = emptyMap(),
        refreshUrl: String? = null,
        onAuthRawHandle: Long? = null
    ): Auth = Auth(
        accessToken = accessToken,
        accessHeaders = accessHeaders,
        accessQueryParams = accessQueryParams,
        accessTokenExpiryTime = accessTokenExpiryTime,
        refreshToken = refreshToken,
        refreshHeaders = refreshHeaders,
        refreshQueryParams = refreshQueryParams,
        refreshUrl = refreshUrl,
        onAuthRawHandle = onAuthRawHandle
    )

    @Test
    fun `getExpandedAccessHeaders replaces tokens correctly`() {
        val auth = createAuth(
            accessToken = "access123",
            refreshToken = "refresh123",
            accessHeaders = mapOf("Authorization" to "Bearer {accessToken}", "Refresh" to "{refreshToken}")
        )
        val expandedHeaders = auth.getExpandedAccessHeaders()
        assertEquals("Bearer access123", expandedHeaders["Authorization"])
        assertEquals("refresh123", expandedHeaders["Refresh"])
    }

    @Test
    fun `getExpandedAccessQueryParams replaces tokens correctly`() {
        val auth = createAuth(
            accessToken = "access456",
            refreshToken = "refresh456",
            accessQueryParams = mapOf("token" to "A {accessToken}", "refresh" to "A {refreshToken}")
        )
        val expandedParams = auth.getExpandedAccessQueryParams()
        assertEquals("A access456", expandedParams["token"])
        assertEquals("A refresh456", expandedParams["refresh"])
    }

    @Test
    fun `addOrUpdateQueryParams adds new query params to URL`() {
        // ignored because we would have to mock Uri.parse
//        val auth = createAuth()
//        val url = "https://example.com/api"
//        val queryParams = mapOf("param1" to "value1", "param2" to "value2")
//        val updatedUri = auth.addOrUpdateQueryParams(url, queryParams)
//        assertEquals("https://example.com/api?param1=value1&param2=value2", updatedUri.toString())
    }

    @Test
    fun `addOrUpdateQueryParams updates existing query params`() {
        // ignored because we would have to mock Uri.parse
//        val auth = createAuth()
//        val url = "https://example.com/api?param1=oldValue"
//        val queryParams = mapOf("param1" to "newValue")
//        val updatedUri = auth.addOrUpdateQueryParams(url, queryParams)
//        assertEquals("https://example.com/api?param1=newValue", updatedUri.toString())
    }

    @Test
    fun `isTokenExpired returns true when token is expired`() {
        val expiryTime = System.currentTimeMillis() - 10000  // Expired 10 seconds ago
        val auth = createAuth(accessTokenExpiryTime = expiryTime)
        assertTrue(auth.isTokenExpired())
    }

    @Test
    fun `isTokenExpired returns false when token is not expired`() {
        val expiryTime = System.currentTimeMillis() + 60000  // Expires in 1 minute
        val auth = createAuth(accessTokenExpiryTime = expiryTime)
        assertFalse(auth.isTokenExpired())
    }

    @Test
    fun `isTokenExpired applies buffer time correctly`() {
        val expiryTime = System.currentTimeMillis() + 5000  // Expires in 5 seconds
        val auth = createAuth(accessTokenExpiryTime = expiryTime)
        assertTrue(auth.isTokenExpired(bufferTime = 7000))  // Buffer time makes it appear expired
        assertFalse(auth.isTokenExpired(bufferTime = 3000)) // Within buffer, not expired
    }

    @Test
    fun `hasOnAuthCallback returns true when onAuthRawHandle is set`() {
        val auth = createAuth(onAuthRawHandle = 1234L)
        assertTrue(auth.hasOnAuthCallback())
    }

    @Test
    fun `hasOnAuthCallback returns false when onAuthRawHandle is not set`() {
        val auth = createAuth(onAuthRawHandle = null)
        assertFalse(auth.hasOnAuthCallback())
    }
}

File: test/auth_test.dart
import 'dart:convert';
import 'dart:io';
import 'package:background_downloader/background_downloader.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:mockito/annotations.dart';
import 'package:mockito/mockito.dart';
import 'package:http/http.dart' as http;

import 'auth_test.mocks.dart';

@GenerateMocks([http.Client])
void main() {
  group('Auth class functionality', () {
    late MockClient mockClient;
    late Auth auth;

    setUp(() {
      mockClient = MockClient();
      auth = Auth(
        accessToken: 'initialAccessToken',
        refreshToken: 'initialRefreshToken',
        refreshUrl: 'https://example.com/refresh',
        accessTokenExpiryTime: DateTime.now()
            .subtract(const Duration(seconds: 10)), // expired token
      );
    });

    test('No change if token not expired', () async {
      // Set up mock response for refresh request
      when(mockClient.post(
        Uri.parse(auth.refreshUrl!),
        headers: auth.refreshHeaders,
        body: jsonEncode({'refresh_token': auth.refreshToken}),
      )).thenAnswer((_) async => http.Response(
          jsonEncode({'access_token': 'newAccessToken', 'expires_in': 3600}),
          200));
      auth.accessTokenExpiryTime = null; // never expires
      auth.accessQueryParams = {'accessToken': '{accessToken}'};
      Uri uri = await auth.getAccessUri(
        url: 'https://example.com/resource',
        httpClient: mockClient,
      );
      // Verify that the URI includes original query parameters
      expect(uri.queryParameters['accessToken'], 'initialAccessToken');
      expect(auth.accessToken, 'initialAccessToken');
    });

    test('HttpException if client returns error', () async {
      // Set up mock response for refresh request
      when(mockClient.post(
        Uri.parse(auth.refreshUrl!),
        headers: auth.refreshHeaders,
        body: jsonEncode({'refresh_token': auth.refreshToken}),
      )).thenAnswer((_) async => http.Response('', 400));
      auth.accessQueryParams = {'accessToken': '{accessToken}'};
      expect(
          () => auth.getAccessUri(
                url: 'https://example.com/resource',
                httpClient: mockClient,
              ),
          throwsA(const HttpException('Could not refresh access token')));
    });

    test('Token refresh updates accessToken and accessTokenExpiryTime',
        () async {
      // Set up mock response for refresh request
      when(mockClient.post(
        Uri.parse(auth.refreshUrl!),
        headers: {'Content-type': 'application/json'},
        body: jsonEncode({
          'grant_type': 'refresh_token',
          'refresh_token': auth.refreshToken
        }),
      )).thenAnswer((_) async => http.Response(
          jsonEncode({
            'access_token': 'newAccessToken',
            'expires_in': 3600,
            'refresh_token': 'newRefreshToken'
          }),
          200));
      final (updatedAccessToken, updatedRefreshToken) =
          await auth.refreshAccessToken(
        httpClient: mockClient,
      );
      // Check if tokens and expiry time are updated
      expect(updatedAccessToken, true);
      expect(updatedRefreshToken, true);
      expect(auth.accessToken, 'newAccessToken');
      expect(auth.refreshToken, 'newRefreshToken');
      expect(auth.accessTokenExpiryTime!.isAfter(DateTime.now()), true);
    });

    test('getAccessUri refreshes token if expired', () async {
      // Set up mock response for refresh request
      when(mockClient.post(
        Uri.parse(auth.refreshUrl!),
        headers: {'Content-type': 'application/json'},
        body: jsonEncode({
          'grant_type': 'refresh_token',
          'refresh_token': auth.refreshToken
        }),
      )).thenAnswer((_) async => http.Response(
          jsonEncode({'access_token': 'newAccessToken', 'expires_in': 3600}),
          200));
      auth.accessQueryParams = {'accessToken': '{accessToken}'};
      Uri uri = await auth.getAccessUri(
        url: 'https://example.com/resource',
        httpClient: mockClient,
      );
      // Verify that the URI includes updated query parameters
      expect(uri.queryParameters['accessToken'], 'newAccessToken');
      expect(auth.accessToken, 'newAccessToken');
    });

    test('getAccessHeaders returns headers with updated accessToken', () {
      auth.accessToken = 'updatedAccessToken';
      auth.accessHeaders = {'Authorization': 'Bearer {accessToken}'};
      final headers = auth.getAccessHeaders();
      // Check that the accessToken is replaced in the headers
      expect(headers['Authorization'], 'Bearer updatedAccessToken');
    });

    test('isTokenExpired returns true for expired token', () {
      auth.accessTokenExpiryTime =
          DateTime.now().subtract(const Duration(seconds: 1));
      expect(auth.isTokenExpired(), true);
    });

    test('isTokenExpired returns false for non-expired token', () {
      auth.accessTokenExpiryTime =
          DateTime.now().add(const Duration(seconds: 3600));
      expect(auth.isTokenExpired(), false);
    });

    test('getAccessQueryParams includes access token in query parameters', () {
      auth.accessToken = 'testAccessToken';
      auth.accessQueryParams = {'token': '{accessToken}'};
      final queryParams = auth.getAccessQueryParams();
      // Check if the accessToken is correctly included
      expect(queryParams['token'], 'testAccessToken');
    });
  });

  group('Auth JSON serialization tests', () {
    test('toJson returns correct JSON map', () {
      final auth = Auth(
        accessToken: 'testAccessToken',
        accessHeaders: {'Authorization': 'Bearer {accessToken}'},
        accessQueryParams: {'token': '{accessToken}'},
        accessTokenExpiryTime:
            DateTime.fromMillisecondsSinceEpoch(1672531199000),
        refreshToken: 'testRefreshToken',
        refreshUrl: 'https://example.com/refresh',
        refreshHeaders: {'Content-Type': 'application/json'},
        refreshQueryParams: {'grant_type': 'refresh_token'},
      );
      final json = auth.toJson();
      // Check JSON structure
      expect(json['accessToken'], 'testAccessToken');
      expect(json['accessHeaders'], {'Authorization': 'Bearer {accessToken}'});
      expect(json['accessQueryParams'], {'token': '{accessToken}'});
      expect(json['accessTokenExpiryTime'],
          1672531199000); // Milliseconds since epoch
      expect(json['refreshToken'], 'testRefreshToken');
      expect(json['refreshUrl'], 'https://example.com/refresh');
      expect(json['refreshHeaders'], {'Content-Type': 'application/json'});
      expect(json['refreshQueryParams'], {'grant_type': 'refresh_token'});
    });

    test('fromJson creates an Auth object from JSON map', () {
      final json = {
        'accessToken': 'testAccessToken',
        'accessHeaders': {'Authorization': 'Bearer {accessToken}'},
        'accessQueryParams': {'token': '{accessToken}'},
        'accessTokenExpiryTime': 1672531199000,
        'refreshToken': 'testRefreshToken',
        'refreshUrl': 'https://example.com/refresh',
        'refreshHeaders': {'Content-Type': 'application/json'},
        'refreshQueryParams': {'grant_type': 'refresh_token'},
      };
      final auth = Auth.fromJson(json);
      // Check if fields are correctly populated
      expect(auth.accessToken, 'testAccessToken');
      expect(auth.accessHeaders, {'Authorization': 'Bearer {accessToken}'});
      expect(auth.accessQueryParams, {'token': '{accessToken}'});
      expect(auth.accessTokenExpiryTime,
          DateTime.fromMillisecondsSinceEpoch(1672531199000));
      expect(auth.refreshToken, 'testRefreshToken');
      expect(auth.refreshUrl, 'https://example.com/refresh');
      expect(auth.refreshHeaders, {'Content-Type': 'application/json'});
      expect(auth.refreshQueryParams, {'grant_type': 'refresh_token'});
    });

    test('toJson and fromJson consistency', () {
      final originalAuth = Auth(
        accessToken: 'consistentAccessToken',
        accessHeaders: {'Authorization': 'Bearer {accessToken}'},
        accessQueryParams: {'token': '{accessToken}'},
        accessTokenExpiryTime:
            DateTime.fromMillisecondsSinceEpoch(1672531199000),
        refreshToken: 'consistentRefreshToken',
        refreshUrl: 'https://example.com/consistent_refresh',
        refreshHeaders: {'Content-Type': 'application/json'},
        refreshQueryParams: {'grant_type': 'refresh_token'},
      );
      // Convert to JSON and back to an Auth instance
      final json = originalAuth.toJson();
      final newAuth = Auth.fromJson({
        'accessToken': json['accessToken'],
        'accessHeaders': json['accessHeaders'],
        'accessQueryParams': json['accessQueryParams'],
        'accessTokenExpiryTime': json['accessTokenExpiryTime'],
        'refreshToken': json['refreshToken'],
        'refreshUrl': json['refreshUrl'],
        'refreshHeaders': json['refreshHeaders'],
        'refreshQueryParams': json['refreshQueryParams'],
      });
      // Check if new Auth object matches the original one
      expect(newAuth.accessToken, originalAuth.accessToken);
      expect(newAuth.accessHeaders, originalAuth.accessHeaders);
      expect(newAuth.accessQueryParams, originalAuth.accessQueryParams);
      expect(newAuth.accessTokenExpiryTime, originalAuth.accessTokenExpiryTime);
      expect(newAuth.refreshToken, originalAuth.refreshToken);
      expect(newAuth.refreshUrl, originalAuth.refreshUrl);
      expect(newAuth.refreshHeaders, originalAuth.refreshHeaders);
      expect(newAuth.refreshQueryParams, originalAuth.refreshQueryParams);
    });

    test('Handles null values in nullable properties', () {
      final auth = Auth(
        accessToken: null,
        accessHeaders: {},
        accessQueryParams: {},
        accessTokenExpiryTime: null,
        refreshToken: null,
        refreshUrl: null,
        refreshHeaders: {},
        refreshQueryParams: {},
      );
      final json = auth.toJson();
      // Ensure nullable properties are null in JSON output
      expect(json['accessToken'], null);
      expect(json['accessTokenExpiryTime'], null);
      expect(json['refreshToken'], null);
      expect(json['refreshUrl'], null);
      // Convert back from JSON to Auth and check that fields are still null
      final newAuth = Auth.fromJson({
        'accessToken': json['accessToken'],
        'authHeaders': json['accessHeaders'],
        'accessQueryParams': json['accessQueryParams'],
        'accessTokenExpiryTime': json['accessTokenExpiryTime'],
        'reAuthToken': json['refreshToken'],
        'tokenRefreshUrl': json['refreshUrl'],
        'refreshHeaders': json['refreshHeaders'],
        'refreshQueryParams': json['refreshQueryParams'],
      });
      expect(newAuth.accessToken, null);
      expect(newAuth.accessTokenExpiryTime, null);
      expect(newAuth.refreshToken, null);
      expect(newAuth.refreshUrl, null);
    });
  });
}
